File: javascript/src/agent/fixflow/summary.test.ts
// Tests for the status-aware terminal summary builder (pure, no I/O).
import { describe, expect, it } from 'vitest';

import type { Comparison } from '../../githubapi/client';
import { type SummaryInput, TerminalOutcome, buildSummaryText } from './summary';

function input(overrides: Partial<SummaryInput> = {}): SummaryInput {
  return {
    outcome: TerminalOutcome.Success,
    workflow: 'lint',
    fullRepo: 'acme/api',
    prNumber: 42,
    attempts: 1,
    report: '',
    lastOutput: '',
    timeout: '90m',
    checkName: 'agent-lint-verify',
    changed: { totalCommits: 0, files: [] },
    ...overrides,
  };
}

const changed: Comparison = {
  totalCommits: 2,
  files: [
    { path: 'a.ts', status: 'modified', additions: 1, deletions: 0 },
    { path: 'b.ts', status: 'added', additions: 5, deletions: 0 },
  ],
};

describe('buildSummaryText', () => {
  it('frames a success with the change summary and targeted findings', () => {
    const text = buildSummaryText(
      input({ outcome: TerminalOutcome.Success, attempts: 2, report: 'fix unused import', changed }),
    );
    expect(text).toContain('the lint fix passed CI after 2 attempts.');
    expect(text).toContain('2 commits changed a.ts, b.ts.');
    expect(text).toContain('Targeted: fix unused import');
  });

  it('uses the singular attempt phrasing', () => {
    const text = buildSummaryText(input({ outcome: TerminalOutcome.Success, attempts: 1 }));
    expect(text).toContain('after 1 attempt.');
  });

  it('frames an exhausted run with the remaining findings', () => {
    const text = buildSummaryText(
      input({ outcome: TerminalOutcome.Exhausted, attempts: 3, lastOutput: 'still failing: x' }),
    );
    expect(text).toContain('the lint fix still fails CI after 3 attempts. Please review.');
    expect(text).toContain('Remaining: still failing: x');
  });

  it('frames a timeout with the awaited check name and timeout', () => {
    const text = buildSummaryText(
      input({ outcome: TerminalOutcome.Timeout, attempts: 1, report: 'r', timeout: '90m' }),
    );
    expect(text).toContain('saw no CI result after 90m waiting for agent-lint-verify (1 attempt).');
    expect(text).toContain('Please review.');
  });

  it('reports no changes when the comparison is empty', () => {
    const text = buildSummaryText(input({ changed: { totalCommits: 0, files: [] } }));
    expect(text).toContain('No changes were recorded on the PR.');
  });

  it('truncates a long file list with a (+N more) suffix', () => {
    const files = Array.from({ length: 10 }, (_, i) => ({
      path: `f${i}.ts`,
      status: 'modified',
      additions: 1,
      deletions: 0,
    }));
    const text = buildSummaryText(input({ changed: { totalCommits: 1, files } }));
    expect(text).toContain('1 commit changed f0.ts, f1.ts, f2.ts, f3.ts, f4.ts, f5.ts, f6.ts, f7.ts (+2 more).');
  });

  it('collapses whitespace and truncates an overlong findings blob', () => {
    const blob = 'a'.repeat(300) + '\n\n   ' + 'b'.repeat(50);
    const text = buildSummaryText(input({ outcome: TerminalOutcome.Success, report: blob }));
    const line = text.split('\n').find((l) => l.startsWith('Targeted: '))!;
    // 280 runes + the ellipsis, after the "Targeted: " label.
    expect([...line.slice('Targeted: '.length)]).toHaveLength(281);
    expect(line.endsWith('…')).toBe(true);
  });

  it('omits the findings line when the blob is empty', () => {
    const text = buildSummaryText(input({ outcome: TerminalOutcome.Success, report: '' }));
    expect(text).not.toContain('Targeted:');
  });
});

File: kotlin/src/test/kotlin/com/automation/agent/agent/fixflow/SummaryTest.kt
package com.automation.agent.agent.fixflow

import com.automation.agent.githubapi.ChangedFile
import com.automation.agent.githubapi.Comparison
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

private fun changed(commits: Int, vararg files: String): Comparison =
    Comparison(totalCommits = commits, files = files.map { ChangedFile(path = it) })

class SummaryTest : BehaviorSpec({
    Given("a successful outcome") {
        When("building the summary") {
            Then("it frames success with attempts, changes, and the targeted findings") {
                val text = buildSummaryText(
                    SummaryInput(
                        outcome = TerminalOutcome.SUCCESS, workflow = "lint", fullRepo = "acme/api", prNumber = 7,
                        attempts = 2, report = "fix the unused import", changed = changed(1, "a.kt"),
                    ),
                )
                text shouldContain "acme/api: the lint fix passed CI after 2 attempts."
                text shouldContain "1 commit changed a.kt."
                text shouldContain "Targeted: fix the unused import"
            }
        }
    }

    Given("an exhausted outcome") {
        When("building the summary") {
            Then("it asks for review and inlines the remaining findings") {
                val text = buildSummaryText(
                    SummaryInput(
                        outcome = TerminalOutcome.EXHAUSTED, workflow = "coverage", fullRepo = "acme/api", prNumber = 7,
                        attempts = 3, lastOutput = "still 40% covered", changed = changed(3, "x.kt", "y.kt"),
                    ),
                )
                text shouldContain "still fails CI after 3 attempts. Please review."
                text shouldContain "3 commits changed x.kt, y.kt."
                text shouldContain "Remaining: still 40% covered"
            }
        }
    }

    Given("a timeout outcome") {
        When("building the summary") {
            Then("it reports the timeout, the awaited check, and 1-attempt phrasing") {
                val text = buildSummaryText(
                    SummaryInput(
                        outcome = TerminalOutcome.TIMEOUT, workflow = "lint", fullRepo = "acme/api", prNumber = 7,
                        attempts = 1, report = "tidy imports", timeout = "90m", checkName = "agent-lint-verify",
                    ),
                )
                text shouldContain "saw no CI result after 90m waiting for agent-lint-verify (1 attempt)."
                text shouldContain "No changes were recorded on the PR."
                text shouldContain "Targeted: tidy imports"
            }
        }
    }

    Given("more than eight changed files") {
        When("building the summary") {
            Then("the file list is truncated with a +N more suffix") {
                val files = (1..10).map { "f$it.kt" }.toTypedArray()
                val text = buildSummaryText(
                    SummaryInput(outcome = TerminalOutcome.SUCCESS, workflow = "lint", fullRepo = "acme/api", prNumber = 1, attempts = 1, changed = changed(1, *files)),
                )
                text shouldContain "(+2 more)"
                text shouldContain "f8.kt"
                text shouldNotContain "f9.kt"
            }
        }
    }

    Given("an empty findings blob") {
        When("building the summary") {
            Then("no findings line is appended") {
                val text = buildSummaryText(SummaryInput(outcome = TerminalOutcome.SUCCESS, workflow = "lint", fullRepo = "acme/api", prNumber = 1, attempts = 1, report = "   "))
                text shouldNotContain "Targeted:"
            }
        }
    }

    Given("various durations") {
        When("formatting the timeout") {
            Then("it uses the most compact whole unit") {
                formatTimeout(90.minutes) shouldBe "90m"
                formatTimeout(1.hours) shouldBe "1h"
                formatTimeout(30.seconds) shouldBe "30s"
            }
        }
    }
})
